### PR TITLE
allow multiple `Definitions` objects in a single module

### PR DESCRIPTION
**Pull request by sryza**
## Summary & Motivation

We recently introduced `Definitions.merge`, with the aim of enabling this kind of pattern:

```python
from marketing import marketing_defs
from sales import sales_defs

defs = Definitions.merge(marketing_defs, sales_defs)
```

However, the above snippet currently fails with
```
DagsterInvariantViolationError("Cannot have more than one Definitions object defined at module scope")
```

This PR makes it no longer fail, by special-casing variables named `defs`.

### CODE REVIEW

**Review by schrockn — CHANGES_REQUESTED** [review]
Definitely a bit spooky to hardcode a name like this. I wonder if we should take this opportunity to add the new API that returns `Definitions` from a function that is decoratored or something to be more explicit?

The other thing this is sparking in my mind is should we have a different type for bundles of unbound definitions.

If the meantime to unblock yourself you can wrap the imported defs in functions.

**Comment by schrockn** [comment]
I'm not super super opposed to this. Just req'ing changes to spark conversation.

**Comment by sryza** [comment]
> If the meantime to unblock yourself you can wrap the imported defs in functions.

Yeah, not personally blocked on this. But holding off on making `Definitions.merge` public before working this out, because it's a footgun.

> Definitely a bit spooky to hardcode a name like this.

FWIW, just based on my experience as a Dagster user whose fingers can type `defs = Definitions` without hitting main memory, I was actually fairly surprised to discover this name wasn't hardcoded.

> The other thing this is sparking in my mind is should we have a different type for bundles of unbound definitions.

Yeah it seems like, if we don't go the "making 'defs' special" route, we would need to introduce a new type.

Something to think about with respect to a new type is that the following can be orthogonal:
- Signaling that an object is unbound
- Signaling that an object is the main one that should be loaded for the code location

I.e. someone might want to validate that the definitions in a submodule have all their requirements satisfied, then import it in their definitions.py, then merge it with some other stuff.

Linking earlier brainstorm on type name for reference: https://github.com/dagster-io/dagster/pull/21746#issuecomment-2123522385.

**Comment by sryza** [comment]
That doesn't help much with import perf unless they put everything inside that function, right? I think that gets pretty clunky.

foo.py:
```python
@asset
def asset1():
    ...

def defs() -> Definitions:
   return Definitions(assets=[asset1])
```

The code to define asset1 is still going to run.


Separate thing, independent of merging blueprints, is that this helps with applying properties to assets in bulk (which is a big gap in our APIs right now):

```python
raw_defs = Definition(...)
defs_with_tags = raw_defs.map_asset_specs(my_tags_function)
defs = link_code_references_to_git(defs_with_tags)
```

Of course, they could put all of this stuff in one statement, but that can get unwieldy taken to the limit.

**Comment by schrockn** [comment]
It helps with anything that uses load_from or anything programmatic like that. Super slow code paths like loading a dbt project top of mind for me here.

**Comment by sryza** [comment]
I find it scary, but another option to think about here is to introduce `CodeLocation`:

```python
from marketing import marketing_defs
from sales import sales_defs

code_location = CodeLocation(defs=Definitions.merge(marketing_defs, sales_defs))
```

**Comment by shalabhc** [comment]
As a strong proponent of "lets not do heavy compute during import", I like the suggestion to use a function `def defs() -> Definitions:`. Even though we currently evaluate `@asset` during import, it seems possible to change it so that:
1. `@asset` only adds a tag to the function, or adds it to a registry
2. The `Definitions()` constructor does the heavy compute for all included assets 

I'm not saying we change `@asset` right now, but making `defs` a function allows us to moving more compute out of.

**Comment by sryza** [comment]
> What would you think about using this as an opportunity to introduce a new API that can target functions that return Definitiions

Realized I didn't directly answer this question. Despite the concerns I raised about how widely applicable it would be, I feel mostly positive about supporting this. One question is whether we'd require the function to have a `-> Definitions` return type annotation.

**Comment by schrockn** [comment]
I think for now we should just expect users to write functions the return Definitions if they want to merge. I also think we should predicate any of these decisions on dogfooding in dope.

**Comment by sryza** [comment]
> I also think we should predicate any of these decisions on dogfooding in dope.

Dogfooding in DOP was actually how we became aware of the issue that this PR is trying to solve. We have a workaround in DOP, but that footgun-i-ness of that experience is what led me to believe that we shouldn't make `Definitions.merge` public until we improve it.

**Comment by sryza** [comment]
To be clear, I think we should resolve this for 1.8, and the main reason I'm pushing on it now is because I like to frontload decisions so we're not scrambling at the end. But don't think this needs to get merged urgently.

**Comment by sryza** [comment]
Here's a prototype of the approach from [this comment](https://github.com/dagster-io/dagster/pull/22364#issuecomment-2163159624) above: https://github.com/dagster-io/dagster/pull/23253

**Pull request closed without merging**
